### ISSUE TITLE
Sleeping causes the player to roll till the player is tired and then puts them to sleep, regardless of actual tiredness

### ISSUE THREAD

**Issue by Zachert45** [CleverRaven/Cataclysm-DDA#28326]
When going to sleep and confirming to go to sleep, the player will toss and turn as if troubled sleeping and will due this action till such time as tiredness is at a low enough level to fall asleep. This chain of events can be done at any time, even if the player has just woken up from a full night of sleep. This leads the player to "sleep" long amounts of time without any chance to stop the action. Normal actions of waking the player up such as noise, attacks, or alarms gives the player the option to wake up or ignore said item of trying to get them out of the action.

**To Reproduce**  
1. Hit your context key for player sleep, default ($)
2. Allow sleep to progress after confirming in menus
3. Watch as time fly's by without being able to stop. (.) will still allow for you to stop while on tossing phase.

**Expected behavior**  
I am unsure how changes to sleeping were changed, it has been some time since I last played CDDA, after asking on the discord chat about this issue, some of the other have stated that other people have asked about this. The last time I have played CDDA (nine or more months ago) attempting to sleep when not sleepy will prevent the player from sleeping with a cannot sleep message with some minor time passage. If this is a current feature for sleeping and is as intended then I am sorry for opening this issue ticket. If intended as feature, perhaps a lock on how long you will spend attempting to sleep again be added, faster computers or where slow down is not an issue has time flying by if you are not paying attention.

 - OS: Windows 10 (1809 build)
 - Game Version: 0.C-37555-gf76242f
 - Graphics version: Tiles
 - Mods loaded: No mods used in the testing

Note this may be related to Impossible to sleep with insomniac trait on uncomfortable surfaces i,e pine lean-to. #28247

**Comment by FantasticFwoosh** [comment]
'Dead Tired' is immediate sleeping in favourable conditions, other factors like temperature, current level of pain & how well sated you are contribute to the time if not continous "tossing and turning" that occurs which counts as partial sleeping.

I think this is just a balancing issue on a feature that requires to be finely tuned or made more forgiving; when I end up sleeping very successfully via a combination of the positive sleeping perk and a sedentary lifestyle with food stocks and some minor tasks do be done during the day in as ideal condition that i can set up my character drops off to sleep in a short amount of time.

**Comment by Inglonias** [comment]
As of [PR #25612](https://github.com/CleverRaven/Cataclysm-DDA/pull/25612), attempting to fall asleep was made into an activity, rather than an effect. The main reason for this was to fix an issue where zombies could attack your stuff or your allies while you were trying to fall asleep and you couldn't do anything for 30 minutes. Sleeping currently works as follows:

1. If you try to fall asleep, an activity begins.
2. The activity ends when you have woken up after falling asleep, your alarm goes off (regardless of how much you slept, or if 24 hours have passed, whichever comes first.

That part of the issue is working as intended as far as I can tell. Regarding how easy/hard it is to fall asleep in different situations, I can't really comment on that. I rarely have an issue trying to sleep, but nine times out of ten I am sleeping someplace comfortable.

**Comment by kevingranade** [comment]
We could add a check to cancel out of the action if you have no chance to fall asleep.  Everything else seems to be working as intended.

**Comment by Zachert45** [comment]
After playing a bit more, the system works as it was designed to. Adding a check to ask get out of attempting to sleep every so often in case of such of event that you cannot stop normally would be a simple fix, I think my issue was more of coming into a feature for the first time (in this case the new sleep system). As is the problem for people not playing experimental builds for a long time or new players, its a learn by failing first system to know better the next time.

If nothing else this ticket only serves at this point to highlight a feature that may need looking at the future, the focus on 0.D release build is more of the focus now. As always this is only one of many things that need looked at some point, but for things how they work now players just need to manage sleep cycles better and this is not much of an issue. If nothing else is needed to be discussed the ticket can be closed.

**Comment by aawilson** [comment]
Just to note, this can get incredibly pathological--trying to sleep in the cold on a rock floor traps the character in "toss and turn" for _days_, easily achieving famished/dehydrated/dead tired statuses without any chance to cancel.

**Comment by Fris0uman** [comment]
So since trying to fall alseep can be manually interrupted at any time I don't get what is the issue here.

**Comment by anothersimulacrum** [comment]
People who don't know that or aren't paying attention could potentially waste a lot of time trying to fall asleep?
Not sure how much of a concern that it.

**Comment by tinukedaya** [comment]
Maybe do a waiting overlay with hint how to cancel?
'___________________________________________
| Wating to fall asleep (press 5 to cancel) |
¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯

**Comment by paulenka-aleh** [comment]
When sleeping away from the city in a safe place my character manages to turn and toss for about 3-4 hours before I could stop it or he/she falls asleep. I would rather take a pill or work/read until I'm tired if I didn't manage to fall asleep in 30-60 minutes. As of now the only chance to do that is to hit the keyboard really fast.
tldr; having a single prompt to stop or continue in 30-60 minutes would be nice

**Comment by Fris0uman** [comment]
Ok so time might flow too fast for effective manual interruption.

So a prompt `You have trouble sleeping, keep trying? Y/N` every hour?

**Comment by paulenka-aleh** [comment]
If every hour / every 30 mins I'd say with 3 options:
 - Stop trying to fall asleep and get up
 - Continue trying to fall asleep
 - Continue trying to fall asleep *and don't prompt further* to allow falling asleep continously without keeping the hand on the keyboard

**Comment by Fris0uman** [comment]
Sounds reasonnable, I'll see if I can make that happen.
